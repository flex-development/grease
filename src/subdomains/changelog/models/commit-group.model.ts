/**
 * @file Models - CommitGroup
 * @module grease/changelog/models/CommitGroup
 */

import type { ICommitGroup } from '#src/changelog/interfaces'
import type { Commit, ICommit } from '#src/git'
import {
  get,
  select,
  trim,
  type Assign,
  type Simplify
} from '@flex-development/tutils'

/**
 * Commit group data transfer object.
 *
 * @template T - Parsed commit type
 */
type CommitGroupDTO<T extends Commit = Commit> = Assign<
  Partial<ICommitGroup<T>>,
  Pick<ICommitGroup, 'key'>
>

/**
 * Parsed commit group model.
 *
 * @see {@linkcode ICommitGroup}
 *
 * @template T - Parsed commit type
 *
 * @class
 * @implements {ICommitGroup<T>}
 */
class CommitGroup<T extends Commit = Commit> implements ICommitGroup<T> {
  /**
   * Parsed commits.
   *
   * @public
   * @instance
   * @member {T[]} commits
   */
  public commits: T[]

  /**
   * Remove commit group from changelog.
   *
   * @default isEmptyString(section)
   *
   * @public
   * @instance
   * @member {boolean} hidden
   */
  public hidden: boolean

  /**
   * Group key.
   *
   * @public
   * @instance
   * @member {string} key
   */
  public key: string

  /**
   * Group title.
   *
   * @default key
   *
   * @public
   * @instance
   * @member {string} section
   */
  public section: string

  /**
   * Create a new commit group.
   *
   * @see {@linkcode CommitGroupDTO}
   *
   * @param {CommitGroupDTO<T>} opts - Commit group options
   */
  constructor(opts: CommitGroupDTO<T>) {
    this.commits = get(opts, 'commits', [])
    this.key = get(opts, 'key')
    this.section = trim(get(opts, 'section', this.key))
    this.hidden = get(opts, 'hidden', false)

    !this.commits.length && (this.hidden = true)
    !this.section && (this.hidden = true)
  }

  /**
   * Get a JSON-serializable commit group object.
   *
   * @template U - JSON serializable commit type
   *
   * @public
   *
   * @return {Simplify<ICommitGroup<U>>} JSON-serializable commit group
   */
  public toJSON<U extends ICommit = ICommit>(): Simplify<ICommitGroup<U>> {
    return {
      commits: select(this.commits, null, commit => <U>commit.toJSON()),
      hidden: this.hidden,
      key: this.key,
      section: this.section
    }
  }
}

export { CommitGroup as default, type CommitGroupDTO }
