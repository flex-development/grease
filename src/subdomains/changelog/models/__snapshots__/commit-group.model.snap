// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:changelog/models/CommitGroup > #toJSON > should return json-serializable commit group 1`] = `
{
  "commits": [
    {
      "author": {
        "email": "unicornware@flexdevelopment.llc",
        "name": "Lexus Drumgold",
      },
      "body": "",
      "breaking": false,
      "breaks": [],
      "date": "2023-09-12T02:07:00-04:00",
      "hash": "b61cac0",
      "header": "feat(providers): \`GitService\`",
      "mentions": [],
      "references": [],
      "scope": "providers",
      "sha": "b61cac0a01f3d6d78e05b410bb4008842ffe16c4",
      "subject": "\`GitService\`",
      "tags": [],
      "trailers": [
        {
          "token": "Signed-off-by",
          "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
        },
      ],
      "type": "feat",
      "version": null,
    },
    {
      "author": {
        "email": "unicornware@flexdevelopment.llc",
        "name": "Lexus Drumgold",
      },
      "body": "",
      "breaking": false,
      "breaks": [],
      "date": "2023-09-08T02:35:24-04:00",
      "hash": "094f1b2",
      "header": "feat(cli): \`BumpCommand\`",
      "mentions": [],
      "references": [],
      "scope": "cli",
      "sha": "094f1b2d3c28556225d4cb9265ad7a032115e21f",
      "subject": "\`BumpCommand\`",
      "tags": [],
      "trailers": [
        {
          "token": "Signed-off-by",
          "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
        },
      ],
      "type": "feat",
      "version": null,
    },
    {
      "author": {
        "email": "unicornware@flexdevelopment.llc",
        "name": "Lexus Drumgold",
      },
      "body": "",
      "breaking": false,
      "breaks": [],
      "date": "2023-09-08T01:40:30-04:00",
      "hash": "17c153f",
      "header": "feat(providers): \`BumpService\`",
      "mentions": [],
      "references": [],
      "scope": "providers",
      "sha": "17c153f409200d10e6e40f66826740c16208e63e",
      "subject": "\`BumpService\`",
      "tags": [],
      "trailers": [
        {
          "token": "Signed-off-by",
          "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
        },
      ],
      "type": "feat",
      "version": null,
    },
    {
      "author": {
        "email": "unicornware@flexdevelopment.llc",
        "name": "Lexus Drumgold",
      },
      "body": "",
      "breaking": false,
      "breaks": [],
      "date": "2023-09-05T02:03:28-04:00",
      "hash": "bfb3642",
      "header": "feat(providers): \`PackageService\`",
      "mentions": [],
      "references": [],
      "scope": "providers",
      "sha": "bfb364275b741d4892afcc438a7761b46115f34b",
      "subject": "\`PackageService\`",
      "tags": [],
      "trailers": [
        {
          "token": "Signed-off-by",
          "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
        },
      ],
      "type": "feat",
      "version": null,
    },
    {
      "author": {
        "email": "unicornware@flexdevelopment.llc",
        "name": "Lexus Drumgold",
      },
      "body": "",
      "breaking": false,
      "breaks": [],
      "date": "2023-09-04T20:42:13-04:00",
      "hash": "47d4d7d",
      "header": "feat(decorators): \`IsManifestId\`, \`IsReleaseVersion\`",
      "mentions": [],
      "references": [],
      "scope": "decorators",
      "sha": "47d4d7df5fc91193a28a3d2b5904663a2890a0d0",
      "subject": "\`IsManifestId\`, \`IsReleaseVersion\`",
      "tags": [],
      "trailers": [
        {
          "token": "Signed-off-by",
          "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
        },
      ],
      "type": "feat",
      "version": null,
    },
    {
      "author": {
        "email": "unicornware@flexdevelopment.llc",
        "name": "Lexus Drumgold",
      },
      "body": "",
      "breaking": false,
      "breaks": [],
      "date": "2023-09-04T16:33:20-04:00",
      "hash": "208898c",
      "header": "feat(decorators): \`IsNilable\`, \`IsNullable\`, \`IsOptional\`",
      "mentions": [],
      "references": [],
      "scope": "decorators",
      "sha": "208898c9102c2ab64cf167ed4a391264a5a78ba2",
      "subject": "\`IsNilable\`, \`IsNullable\`, \`IsOptional\`",
      "tags": [],
      "trailers": [
        {
          "token": "Signed-off-by",
          "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
        },
      ],
      "type": "feat",
      "version": null,
    },
    {
      "author": {
        "email": "unicornware@flexdevelopment.llc",
        "name": "Lexus Drumgold",
      },
      "body": "",
      "breaking": false,
      "breaks": [],
      "date": "2023-09-04T16:26:30-04:00",
      "hash": "812e10c",
      "header": "feat(models): \`Version\`",
      "mentions": [],
      "references": [],
      "scope": "models",
      "sha": "812e10c3a5054733373baeeca6a9134a89a478a8",
      "subject": "\`Version\`",
      "tags": [],
      "trailers": [
        {
          "token": "Signed-off-by",
          "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
        },
      ],
      "type": "feat",
      "version": null,
    },
  ],
  "hidden": false,
  "key": "feat",
  "section": ":sparkles: Features",
}
`;
